File: service_pro/public/js/sales_invoice.js

cur_frm.cscript.incentives = function(frm){
    compute_incentives(cur_frm)
}
cur_frm.cscript.sales_team_remove = function(frm){
    compute_incentives(cur_frm)
}
cur_frm.cscript.sales_person = function(frm,cdt, cdn){
    var d = locals[cdt][cdn]
    if(d.sales_person){
        frappe.db.get_value('Sales Person', d.sales_person, "liabilities_account")
        .then(liabilities_account => {
            console.log(liabilities_account)
            cur_frm.doc.liabilities_account = liabilities_account.message.liabilities_account
        cur_frm.refresh_field("liabilities_account")
        })
    }
}
var defaults = {}
frappe.ui.form.on('Sales Invoice', {
    company: function () {
       if(cur_frm.doc.company){
            frappe.call({
                method: "service_pro.doc_events.utils.get_production_settings_defaults",
                args: {
                    company: cur_frm.doc.company
                },
                freeze: true,
                freeze_message: "Getting Company Defaults....",
                callback: function (r) {
                    defaults = r.message

                }
            })
        }
    },
   
    onload: function (frm) {
        console.log("ONLOAD")
        if(cur_frm.doc.company) {
            frappe.call({
                method: "service_pro.doc_events.utils.get_production_settings_defaults",
                args: {
                    company: cur_frm.doc.company
                },
                freeze: true,
                freeze_message: "Getting Company Defaults....",
                callback: function (r) {
                    defaults = r.message

                }
            })
        }
    },
    refresh: function(frm) {
        // frm.toggle_display("update_stock", frm.doc.is_pos)
        frappe.db.get_single_value("Production Settings", "credit_note_user_role").then((value) => {
            if(value) {
                if (!frappe.user_roles.includes(value)) {
                    frm.toggle_display("is_return", false)
                    setTimeout(() => {
                            frm.remove_custom_button('Return / Credit Note', "Create");
                    }, 500);
                }
            }                           
        });
        frm.set_query('cost_center', function() {
            if (frm.doc.company) {
                return {
                    filters: {
                        company: frm.doc.company,
                        is_group: 0,
                    }
                };
            }
        });
        frm.set_query('set_warehouse', function() {
            if (frm.doc.company) {
                return {
                    filters: {
                        company: frm.doc.company
                    }
                };
            }
        });  
    },
    is_pos: function(frm) {
        // frm.toggle_display("update_stock", frm.doc.is_pos)
        frm.doc.update_stock = 1
        if (!frm.doc.is_pos){
            frm.doc.update_stock = 0
        }
    },

    customer: function(frm){

        filter_link_field(cur_frm)
        if(cur_frm.doc.customer){
            frappe.db.get_doc("Customer", cur_frm.doc.customer)
                .then(customer => {
                    if(customer.sales_team.length > 0){
                         cur_frm.doc.sales_man = customer.sales_team[0].sales_person
                         cur_frm.doc.sales_partner = customer.sales_team[0].sales_person
                        cur_frm.refresh_fields(["sales_man","sales_partner"])
                    }

            })
        }
    },
    onload_post_render: function () {
        if(cur_frm.is_new()){
            cur_frm.doc.sales_partner = ""
            cur_frm.doc.incentive = 0
            cur_frm.doc.paid = 0
            cur_frm.doc.unpaid = 0
            cur_frm.refresh_fields(["sales_partner","incentive","paid","unpaid"])
        }
    }
})
cur_frm.cscript.paid = function(frm){
    cur_frm.doc.expense_cost_center = defaults['sales_partner_payments_details'].cost_center
    cur_frm.refresh_field("expense_cost_center")

    // cur_frm.doc.expense_account = defaults['sales_partner_payments_details'].expense_accounts
    // cur_frm.refresh_field("expense_account")

    cur_frm.doc.showroom_cash = defaults['sales_partner_payments_details'].showroom_cash
    cur_frm.refresh_field("showroom_cash")

    if(cur_frm.doc.paid){
        cur_frm.doc.unpaid = 0
        cur_frm.doc.cash = 1

        cur_frm.refresh_field("unpaid")
        cur_frm.refresh_field("cash")
    }


}
cur_frm.cscript.unpaid = function(frm){
    // frappe.db.get_single_value('Production Settings', 'expense_cost_center')
    //     .then(expense_cost_center => {
    cur_frm.doc.expense_cost_center = defaults['sales_partner_payments_details'].cost_center
    cur_frm.refresh_field("expense_cost_center")
        // })
    // frappe.db.get_single_value('Production Settings', 'expense_account')
    //         .then(expense_account => {
    //             console.log(expense_account)
    // cur_frm.doc.expense_account = defaults['sales_partner_payments_details'].expense_accounts
    // cur_frm.refresh_field("expense_account")
    cur_frm.doc.liabilities_account = defaults['sales_partner_payments_details'].expense_accounts
    cur_frm.refresh_field("liabilities_account")
        // })
   if(cur_frm.doc.unpaid){
        cur_frm.doc.paid = 0
        cur_frm.doc.cash = 0
        cur_frm.refresh_field("paid")
        cur_frm.refresh_field("cash")
   }


}
cur_frm.cscript.onload = function(frm){
    if(cur_frm.doc.docstatus === 0 && cur_frm.doc.production.length > 0 && cur_frm.doc.taxes_and_charges){
        cur_frm.trigger("taxes_and_charges")
        cur_frm.save()
    }
    if(cur_frm.is_new() && cur_frm.doc.is_return){
        for(var i =0; i < cur_frm.doc.production.length;i+=1){
            cur_frm.doc.production[i].qty = 0 - cur_frm.doc.production[i].qty
            cur_frm.refresh_field("production")
        }
    }
    filter_link_field(cur_frm)

    cur_frm.set_query('expense_cost_center', () => {
        return {
            filters: {
                is_group: 0,
            }
        }
    });
    if('sales_partner_payments_details' in defaults){
        // cur_frm.doc.expense_account = defaults['sales_partner_payments_details'].expense_account
        // cur_frm.refresh_field("expense_account")

        cur_frm.doc.showroom_cash = defaults['sales_partner_payments_details'].showroom_cash
        cur_frm.refresh_field("showroom_cash")

        // cur_frm.doc.showroom_card = defaults['incentive_journal_defaults'].showroom_card
        // cur_frm.refresh_field("showroom_card")
    }
    if('sales_partner_payments_details' in defaults){

        cur_frm.doc.showroom_cash = defaults['sales_partner_payments_details'].showroom_cash
        cur_frm.refresh_field("showroom_cash")
    }


}
function compute_incentives(cur_frm) {
    var incentive_total = 0
    for(var i=0;i<cur_frm.doc.sales_team.length;i+=1){
        incentive_total += cur_frm.doc.sales_team[i].incentives
    }
    cur_frm.doc.incentive = incentive_total
    cur_frm.refresh_field("incentive")
}
function filter_link_field(cur_frm) {
     cur_frm.set_query('reference', 'production', () => {
        return {
            filters: [
                ["customer", "=", cur_frm.doc.customer],
                ["docstatus", "=", 1],
                ["status", "in", ["To Deliver and Bill", "To Bill","Partially Completed", "Partially Delivered"]],
            ]
        }
    })

}

cur_frm.cscript.reference = function (frm,cdt,cdn) {
    var d = locals[cdt][cdn]
    if(d.reference){
         frappe.db.get_doc('Production', d.reference)
            .then(doc => {
                var add = true
                var qty_1 = doc.qty
                for(var x=0;x < cur_frm.doc.items.length;x += 1){
                    if(cur_frm.doc.items[x].item_code === doc.item_code_prod){
                        add = false
                    }
                }
                frappe.call({
                    method: "service_pro.service_pro.doctype.production.production.get_dn_si_qty",
                    args: {
                        item_code: doc.item_code_prod,
                        qty:   doc.qty,
                        name: d.reference
                    },
                    async: false,
                    callback: function (r) {
                        qty_1 = r.message

                    }
                })
                d.qty = qty_1
                d.rate = doc.invoice_rate
                d.amount =doc.invoice_rate *qty_1
                cur_frm.refresh_field("production")

        if(add){
        // frappe.db.get_single_value('Production Settings', 'income_account')
            // .then(income_account => {
                frappe.db.get_doc('Item', doc.item_code_prod)
                    .then(doc1 => {
                         cur_frm.add_child('items', {
                            item_code: doc.item_code_prod,
                            qty: qty_1,
                            uom: doc.umo,
                            stock_uom: doc.umo,
                            rate: doc.invoice_rate,
                            amount: doc.invoice_rate * qty_1,
                            item_name: doc1.item_name,
                            description: doc1.description,
                            income_account: defaults['finish_good_defaults'].income_account,

                        });

                    cur_frm.refresh_field('items');
                })
            // })

            if(doc.sales_man){
                cur_frm.add_child('sales_man', {
                    sales_man: doc.sales_man,
                    reference: d.reference,
                });
                cur_frm.refresh_field('sales_man');

            }



        }

            })
    }
}

File: service_pro/doc_events/sales_invoice.py
import frappe
from frappe import _

def generate_jv(doc):
	if doc.paid:
		data = frappe.db.sql(""" SELECT * FROM `tabSales Partner Payments Details` WHERE company=%s """, doc.company,
							 as_dict=1)

		doc_jv = {
			"doctype": "Sales Partner Payments",
			"sales_partner_name": doc.sales_partner,
			"company": doc.company,
			"incentive": doc.incentive,
			"balance_amount": doc.incentive,
			"posting_date": doc.posting_date,
			"status": "Paid",
			"sales_invoice_reference": doc.name,
			"invoice_net_amount": doc.total,
			"customer_id": doc.customer,
			"customer_name": doc.customer_name,
			"cost_center": doc.cost_center,
		}
		if len(data) == 0:
			frappe.throw("Please set Sales Partner Payments Details defaults in Production Settings")
		if len(data) > 0 and not data[0].payable_account:
			frappe.throw("Please set Payable Account in Sales Partner Payments Details defaults in Production Settings")

		if len(data) > 0 and not data[0].expense_accounts:
			frappe.throw("Please set Expense Account in Sales Partner Payments Details defaults in Production Settings")

		if not doc.showroom_cash and not data[0].showroom_cash:
			frappe.throw("Please set Showroom Cash in Production Settings")

		mop_cash = frappe.db.sql(""" SELECT * FROM `tabMode of Payment Account` WHERE parent=%s and company=%s""", (doc.showroom_cash or data[0].showroom_cash,doc.company),
								 as_dict=1)

		if len(mop_cash) > 0:
			doc_jv['payable_account'] = mop_cash[0].default_account
		else:
			frappe.throw("Please add account for company " + doc.company + " in Mode of Payment " + doc.showroom_cash or data[0].showroom_cash)
		doc_jv['expense_account'] = data[0].expense_accounts
		jv = frappe.get_doc(doc_jv)
		jv.insert(ignore_permissions=1)
		jv.submit()
		frappe.msgprint("Sales Partner Payments Created")
	elif doc.unpaid:
		data = frappe.db.sql(""" SELECT * FROM `tabSales Partner Payments Details` WHERE company=%s """,doc.company, as_dict=1)

		doc_jv = {
			"doctype": "Sales Partner Payments",
			"company": doc.company,
			"sales_partner_name": doc.sales_partner,
			"incentive": doc.incentive,
			"balance_amount": doc.incentive,
			"posting_date": doc.posting_date,
			"status": "Unpaid",
			"sales_invoice_reference": doc.name,
			"invoice_net_amount": doc.total,
			"customer_id": doc.customer,
			"customer_name": doc.customer_name,
			"cost_center": doc.cost_center,
		}
		if len(data) == 0:
			frappe.throw("Please set Sales Partner Payments Details defaults in Production Settings")
		if len(data) > 0 and not data[0].payable_account:
			frappe.throw("Please set Payable Account in Sales Partner Payments Details defaults in Production Settings")

		if len(data) > 0 and not data[0].expense_accounts:
			frappe.throw("Please set Expense Account in Sales Partner Payments Details defaults in Production Settings")

		doc_jv['payable_account'] = data[0].payable_account
		doc_jv['expense_account'] = data[0].expense_accounts
		jv = frappe.get_doc(doc_jv)
		jv.insert(ignore_permissions=1)
		jv.submit()
		frappe.msgprint("Sales Partner Payments Created")
		# doc.journal_entry = jv.name
		# frappe.db.sql(""" UPDATE `tabSales Invoice` SET journal_entry=%s WHERE name=%s""", (jv.name, doc.name))
		# frappe.db.commit()

def jv_accounts_unpaid(doc):
	data = frappe.db.sql(""" SELECT * FROM `tabSales Partner Payments` WHERE company=%s """, doc.company, as_dict=1)
	if len(data) == 0:
		frappe.throw("Please setup Sales Partner Payments for company " + doc.company)
	accounts = []
	accounts.append({
		'account': data[0].expense_accounts,
		'debit_in_account_currency': doc.incentive,
		'credit_in_account_currency': 0,
		'cost_center': doc.expense_cost_center,
	})
	accounts.append({
		'account':data[0].payable_account,
		'debit_in_account_currency': 0,
		'credit_in_account_currency': doc.incentive
	})
	return accounts

def jv_accounts_paid(doc):
	data = frappe.db.sql(""" SELECT * FROM `tabSales Partner Payments` WHERE company=%s """, doc.company,as_dict=1)
	if len(data) == 0:
		frappe.throw("Please setup Sales Partner Payments for company " + doc.company)
	accounts = []
	accounts.append({
		'account': data[0].expense_accounts,
		'debit_in_account_currency': doc.incentive,
		'credit_in_account_currency': 0,
		'cost_center': doc.expense_cost_center,
	})
	if doc.cash:
		if not doc.showroom_cash and not data[0].showroom_cash:
			frappe.throw("Please set Showroom Cash in Production Settings")

		mop_cash = frappe.db.sql(""" SELECT * FROM `tabMode of Payment Account` WHERE parent=%s and company=%s """, (doc.showroom_cash or data[0].showroom_cash,doc.company), as_dict=1)
		if len(mop_cash) > 0:
			accounts.append({
				'account': mop_cash[0].default_account,
				'debit_in_account_currency': 0,
				'credit_in_account_currency': doc.incentive
			})
		else:
			frappe.throw("Please add account for company " + doc.company + " in Mode of Payment " + doc.showroom_cash or data[0].showroom_cash)
	else:
		accounts.append({
			'account': data[0].payable_account,
			'debit_in_account_currency': 0,
			'credit_in_account_currency': doc.incentive
		})
	return accounts

def validate_margin_rate_with_rate(doc, method):
	"""Validate that rate field is not lower than custom_margin_rate for each item"""
	if not doc.items:
		return
    
	if frappe.session.user == "Administrator":
		return

	user_roles = frappe.get_roles(frappe.session.user)
	if "Margin Rate Approver" in user_roles:
		return
 
	
	validation_errors = []
	
	for idx, item in enumerate(doc.items):
		if hasattr(item, 'custom_margin_rate') and item.custom_margin_rate and item.rate:
			# Convert to float for comparison
			margin_rate = float(item.custom_margin_rate)
			item_rate = float(item.rate)
			
			# Check if item rate is lower than margin rate
			# Allow submission when rate >= margin_rate
			if item_rate < margin_rate:
				validation_errors.append({
					'row': idx + 1,
					'item_code': item.item_code,
					'item_name': item.item_name or item.item_code,
					'margin_rate': margin_rate,
					'item_rate': item_rate
				})
	
	if validation_errors:
		error_message = "Rate cannot be lower than Margin Rate for the following items:\n\n"
		
		for error in validation_errors:
			error_message += f"Row {error['row']}: ({error['item_code']})\n"
			error_message += f"  • Margin Rate: {error['margin_rate']:.2f}\n"
			error_message += f"  • Item Rate: {error['item_rate']:.2f}\n"
			error_message += f"  • Please increase Item Rate to at least {error['margin_rate']:.2f}\n\n"
		
		frappe.throw(_(error_message), title=_("Rate Below Margin Rate"))

def set_margin_rate_for_document_creator(doc):
	"""Set custom_margin_rate for new documents or update for document creator when their percentage changes"""
	if not doc.company:
		return
		
	current_user = frappe.session.user
	
	if doc.is_new() or doc.owner == current_user:
		user_margin_percentage = get_user_margin_percentage(current_user, doc.company)
		
		if user_margin_percentage is None:
			frappe.logger().info(f"No margin percentage found for user {current_user} in company {doc.company}")
			return
		
		should_update = False
		
		if doc.is_new():
			should_update = True
			frappe.logger().info(f"Setting margin rate {user_margin_percentage} for new document by {current_user}")
		elif doc.owner == current_user and doc.custom_margin_rate != user_margin_percentage:
			should_update = True
			frappe.logger().info(f"Updating margin rate from {doc.custom_margin_rate} to {user_margin_percentage} for document creator {current_user}")
		
		if should_update:
			doc.custom_margin_rate = user_margin_percentage
			
			for item in doc.items:
				if item.item_code and item.warehouse:
					calculate_item_margin_rate(doc, item)

def set_margin_rate_on_load(doc, method):
	"""Set margin rate based on user configuration only for document creator or new documents"""
	set_margin_rate_for_document_creator(doc)

@frappe.whitelist()
def on_submit_si(doc, method):
	if doc.owner == frappe.session.user:
		set_margin_rate_for_document_creator(doc)
	
	# Validate margin rate matches item rate before submission
	validate_margin_rate_with_rate(doc, method)
	
	if doc.selling_price_list:
		if frappe.db.exists("Maximum User Discount", {"parent":doc.selling_price_list, "user": frappe.session.user}):
			max_disc = frappe.db.get_value("Maximum User Discount", {"parent":doc.selling_price_list, "user": frappe.session.user}, ["max_discount"])
			over_disc = frappe.db.exists("Sales Invoice Item", {"parent":doc.name, "discount_percentage": [">", max_disc]}, ['item_code'])
			if max_disc and over_disc:
				frappe.throw(" Maximum allowed discount is {0} for item {1}".format(max_disc, frappe.db.get_value("Sales Invoice Item", over_disc, ['item_code'])))

	if doc.sales_partner and not doc.paid and not doc.unpaid:
		frappe.throw("Please select Paid or Unpaid for Sales Person")

	generate_jv(doc)
	for prod in doc.production:
		production = frappe.db.sql(""" SELECT * FROM `tabProduction` WHERE name=%s """, prod.reference, as_dict=1)
		if len(production) > 0:
			if doc.update_stock and get_dn_si_qty("", production[0].qty, prod.reference) > 0 :
				frappe.db.sql(""" UPDATE `tabProduction` SET status=%s WHERE name=%s""",
							  ("Partially Delivered", prod.reference))
				frappe.db.commit()

			elif get_dn_si_qty("", production[0].qty, prod.reference) == 0 and get_lengths(prod.reference)[0] == get_lengths(prod.reference)[1] :

				frappe.db.sql(""" UPDATE `tabProduction` SET status=%s WHERE name=%s""", ("Completed", prod.reference))
				frappe.db.commit()
				get_service_records(prod.reference)

			elif get_dn_qty(prod.reference) >= 0 and \
                    ((get_dn_si_qty("", production[0].qty, prod.reference) >= 0 and get_lengths(prod.reference)[0] !=
                             get_lengths(prod.reference)[1])):
				frappe.db.sql(""" UPDATE `tabProduction` SET status=%s WHERE name=%s""", ("To Deliver", prod.reference))
				frappe.db.commit()

			elif get_dn_si_qty("", production[0].qty, prod.reference) > 0:
				frappe.db.sql(""" UPDATE `tabProduction` SET status=%s WHERE name=%s""",
							  ("Partially Delivered", prod.reference))
				frappe.db.commit()

def before_save_si(doc, method):
	"""Set margin rate only for document creator or new documents"""
	set_margin_rate_for_document_creator(doc)

def get_service_records(reference):
	estimation_ = ""
	estimation = frappe.db.sql(""" SELECT * FROM `tabProduction` WHERE name= %s""", reference, as_dict=1)
	if len(estimation) > 0:
		estimation_ = estimation[0].estimation
		frappe.db.sql(""" UPDATE `tabEstimation` SET status=%s WHERE name=%s""",
					  ("Completed", estimation_))

	inspections = frappe.db.sql(""" SELECT * FROM `tabInspection Table` WHERE parent=%s """, estimation_, as_dict=1)
	for i in inspections:
		frappe.db.sql(""" UPDATE `tabInspection` SET status=%s WHERE name=%s""",
					  ("Completed", i.inspection))

	srn = frappe.db.sql(""" SELECT * FROM `tabEstimation` WHERE name=%s """, estimation_, as_dict=1)
	if len(srn) > 0:
		srn_ = srn[0].receipt_note
		frappe.db.sql(""" UPDATE `tabService Receipt Note` SET status=%s WHERE name=%s""",
					 ("Completed", srn_))
	frappe.db.commit()


def get_service_records_cancel(reference):
	estimation_ = ""
	estimation = frappe.db.sql(""" SELECT * FROM `tabProduction` WHERE name= %s""", reference, as_dict=1)
	if len(estimation) > 0:
		estimation_ = estimation[0].estimation
		frappe.db.sql(""" UPDATE `tabEstimation` SET status=%s WHERE name=%s""",
					  ("To Production", estimation_))

	inspections = frappe.db.sql(""" SELECT * FROM `tabInspection Table` WHERE parent=%s """, estimation_, as_dict=1)
	for i in inspections:
		frappe.db.sql(""" UPDATE `tabInspection` SET status=%s WHERE name=%s""",
					  ("To Production", i.inspection))

	srn = frappe.db.sql(""" SELECT * FROM `tabEstimation` WHERE name=%s """, estimation_, as_dict=1)
	if len(srn) > 0:
		srn_ = srn[0].receipt_note
		frappe.db.sql(""" UPDATE `tabService Receipt Note` SET status=%s WHERE name=%s""",
					  ("To Production", srn_))
	frappe.db.commit()

@frappe.whitelist()
def on_cancel_si(doc, method):
	for prod in doc.production:
		production = frappe.db.sql(""" SELECT * FROM `tabProduction` WHERE name=%s """, prod.reference, as_dict=1)
		if len(production) > 0:
			if doc.update_stock:
				frappe.db.sql(""" UPDATE `tabProduction` SET status=%s WHERE name=%s""", ("To Deliver and Bill", prod.reference))
				frappe.db.commit()
				get_service_records(prod.reference)

			elif get_lengths(prod.reference)[0] == 0 and get_lengths(prod.reference)[1] == 0:
				frappe.db.sql(""" UPDATE `tabProduction` SET status=%s WHERE name=%s""",
							  ("To Deliver and Bill", prod.reference))
				frappe.db.commit()
				get_service_records(prod.reference)

			elif get_dn_qty(prod.reference) >= 0 and \
					((get_dn_si_qty("", production[0].qty, prod.reference) >= 0 and get_lengths(prod.reference)[0] !=
						get_lengths(prod.reference)[1])):
				frappe.db.sql(""" UPDATE `tabProduction` SET status=%s WHERE name=%s""", ("To Bill", prod.reference))
				frappe.db.commit()

			elif get_dn_si_qty("", production[0].qty, prod.reference) > 0:
				frappe.db.sql(""" UPDATE `tabProduction` SET status=%s WHERE name=%s""",
							  ("Partially Delivered", prod.reference))
				frappe.db.commit()
	if doc.journal_entry:
		frappe.db.sql(""" UPDATE `tabJournal Entry` SET sales_invoice='' WHERE name=%s """, doc.journal_entry)
		frappe.db.commit()

def get_lengths(name):
	si_query = """ 
     			SELECT SIP.qty as qty, SI.status FROM `tabSales Invoice` AS SI 
     			INNER JOIN `tabSales Invoice Production` AS SIP ON SI.name = SIP.parent 
     			WHERE SIP.reference=%s and SIP.parenttype=%s and SI.docstatus = 1 and SI.status!='Cancelled' and SI.update_stock = 0
     			"""
	si = frappe.db.sql(si_query, (name, "Sales Invoice"), as_dict=1)
	dn_query = """ 
    	 			SELECT SIP.qty as qty, DN.status FROM `tabDelivery Note` AS DN 
    	 			INNER JOIN `tabSales Invoice Production` AS SIP ON DN.name = SIP.parent 
    	 			WHERE SIP.reference=%s and SIP.parenttype=%s and DN.docstatus = 1 and DN.status!='Cancelled'
    	 			"""
	dn = frappe.db.sql(dn_query, (name, "Delivery Note"), as_dict=1)

	return len(dn), len(si)

def get_dn_si_qty(item_code, qty, name):
	si_query = """ 
 			SELECT SIP.qty as qty, SI.status FROM `tabSales Invoice` AS SI 
 			INNER JOIN `tabSales Invoice Production` AS SIP ON SI.name = SIP.parent 
 			WHERE SIP.reference=%s and SIP.parenttype=%s and SI.docstatus = 1 and SI.status!='Cancelled'
 			"""
	si = frappe.db.sql(si_query,(name,"Sales Invoice"), as_dict=1)
	dn_query = """ 
	 			SELECT SIP.qty as qty, DN.status FROM `tabDelivery Note` AS DN 
	 			INNER JOIN `tabSales Invoice Production` AS SIP ON DN.name = SIP.parent 
	 			WHERE SIP.reference=%s and SIP.parenttype=%s and DN.docstatus = 1 and DN.status!='Cancelled'
	 			"""
	dn = frappe.db.sql(dn_query,(name, "Delivery Note"), as_dict=1)

	total_qty = 0

	if len(si) > len(dn):
		for i in si:
			total_qty += i.qty

	elif len(dn) > len(si):
		for d in dn:
			total_qty += d.qty
	elif len(dn) == len(si):
		for d in dn:
			total_qty += d.qty
	return float(qty) - float(total_qty)



def get_dn_qty(name):
	dn_query = """ 
	 			SELECT SIP.qty as qty, DN.status FROM `tabDelivery Note` AS DN 
	 			INNER JOIN `tabSales Invoice Production` AS SIP ON DN.name = SIP.parent 
	 			WHERE SIP.reference=%s and SIP.parenttype=%s and DN.docstatus = 1 and DN.status!='Cancelled'
	 			"""
	dn = frappe.db.sql(dn_query, (name, "Delivery Note"), as_dict=1)

	total_qty = 0

	if len(dn) > 0:
		for d in dn:
			total_qty += d.qty

	return float(total_qty)


def validate_so(doc, method):

	if frappe.db.get_single_value("Production Settings", "enable_sales_order_validation"):
		if not doc.is_pos:
			for it in doc.get("items"):
				if not it.sales_order:
					frappe.throw("The Item {0}-{1} Does not have Sales Order".format(it.item_code,it.item_name))

def validate_permission(doc, method):
	if not doc.custom_ignore_permission_ and not doc.custom_production_id:
		frappe.throw("Sales Invoice is Required")

@frappe.whitelist()
def get_role():
	doc = frappe.db.get_value("Production Settings",None,"ignore_permission")
	return doc

def validate_and_calculate_rates(doc, method):
    """Validate and calculate rates based on margin and valuation rate"""
    if doc.custom_margin_rate:
        for item in doc.items:
            if item.item_code and item.warehouse:
                calculate_item_margin_rate(doc, item)


def calculate_item_margin_rate(doc, item):
    """Calculate item margin rate using margin % and valuation rate"""
    try:
        cost_price = frappe.db.get_value("Bin", {
            "item_code": item.item_code,
            "warehouse": item.warehouse
        }, "valuation_rate")

        if not cost_price:
            return  # Skip if no cost price available

        margin_percentage = doc.custom_margin_rate
        margin_decimal = margin_percentage / 100

        if margin_decimal >= 1:
            frappe.throw(f"Margin percentage cannot be 100% or more for item {item.item_code}")

        selling_price = cost_price / (1 - margin_decimal)

        # Set calculated margin price to custom_margin_rate field in child table
        if hasattr(item, 'custom_margin_rate'):
            item.custom_margin_rate = selling_price
        else:
            frappe.logger().warning(f"'custom_margin_rate' field missing in Sales Invoice Item for {item.item_code}")

        frappe.logger().info(f"[Margin Calculation] Item: {item.item_code}, Cost: {cost_price}, "
                             f"Margin: {margin_percentage}%, Rate: {selling_price}")

    except Exception as e:
        frappe.logger().error(f"[Error] Margin calc failed for {item.item_code}: {str(e)}")


def get_user_margin_percentage(user, company):
    """Fetch user-specific margin percentage from Production Settings"""
    try:
        production_settings = frappe.get_doc("Production Settings", company)

        if hasattr(production_settings, 'default_sales_margin_percentage'):
            user_margin = next((row.percentage for row in production_settings.default_sales_margin_percentage
                                if row.user == user), None)
            return user_margin
    except Exception as e:
        frappe.logger().error(f"[Error] Fetching margin for user {user} in {company}: {str(e)}")
    return None


@frappe.whitelist()
def get_user_margin_rate():
    """Return current user's configured margin rate"""
    user = frappe.session.user
    company = frappe.defaults.get_user_default("Company")

    if not company:
        companies = frappe.get_all("Company", filters={"disabled": 0}, fields=["name"], limit=1)
        if companies:
            company = companies[0].name

    return get_user_margin_percentage(user, company)


@frappe.whitelist()
def calculate_selling_price(cost_price, margin_percentage):
    """Calculate selling price given a cost and margin"""
    try:
        cost_price = float(cost_price)
        margin_percentage = float(margin_percentage)

        if cost_price <= 0:
            frappe.throw(_("Cost price must be greater than 0"))

        if not (0 < margin_percentage < 100):
            frappe.throw(_("Margin percentage must be between 0 and 100"))

        margin_decimal = margin_percentage / 100
        selling_price = cost_price / (1 - margin_decimal)

        return {
            "cost_price": cost_price,
            "margin_percentage": margin_percentage,
            "selling_price": selling_price,
            "margin_amount": selling_price - cost_price
        }

    except Exception as e:
        frappe.throw(_("Error calculating selling price: {0}").format(str(e)))


def calculate_item_rate_with_margin(doc, item):
    """[REFERENCE] Compute and update rate field instead of custom_margin_rate"""
    try:
        cost_price = frappe.db.get_value("Bin", {
            "item_code": item.item_code,
            "warehouse": item.warehouse
        }, "valuation_rate")

        if not cost_price:
            return

        margin_percentage = doc.custom_margin_rate
        margin_decimal = margin_percentage / 100

        if margin_decimal >= 1:
            frappe.throw(f"Margin percentage cannot be 100% or more for item {item.item_code}")

        selling_price = cost_price / (1 - margin_decimal)
        item.rate = selling_price  # Reference update

        frappe.logger().info(f"[Ref Margin] Item: {item.item_code}, Cost: {cost_price}, "
                             f"Margin: {margin_percentage}%, Rate: {selling_price}")

    except Exception as e:
        frappe.logger().error(f"[Ref Error] Margin rate set failed for {item.item_code}: {str(e)}")